### ISSUE TITLE
Random add-ons get blocked / tainted after command autocompletion is used

### ISSUE THREAD

**Issue by LenweSaralonde** [Stanzilla/WoWUIBugs#378]
WoW Version : 10.0.2.47213

Random add-ons gets taint error after using command autocompletion in the chat EditBox.

1. Disable all add-ons.
2. Run the following macro to create a custom add-on command: 
```lua
/run SlashCmdList.FOO= function() print("bar") end; SLASH_FOO1 = "/foo"
```
3. Type `/cam` in the chat EditBox then press **Tab** to autocomplete the `/camp` command.
4. Press **Enter**.

Expected result : Character is logged out.
Actual result: "* * * ForceTaint_Strong * * * was blocked" error pop up is shown.

**Comment by Meorawr** [comment]
There's a *very slim* chance this might have been fixed in 10.0.5 already - unfortunately the PTR has been down over the holidays so haven't been able to test, but when they come back up eventually I'll check this.

**Comment by Meorawr** [comment]
> There's a _very slim_ chance this might have been fixed in 10.0.5 already - unfortunately the PTR has been down over the holidays so haven't been able to test, but when they come back up eventually I'll check this.

Just to follow up - the changes I was talking about have no impact, so this remains unfixed in 10.0.5.

**Comment by Meorawr** [comment]
I missed the boat to really confirm this but I _think_ this is the bare minimum fix required:

```diff
diff --git a/Interface/FrameXML/ChatFrame.lua b/Interface/FrameXML/ChatFrame.lua
index ca7bf300..42f8d646 100644
--- a/Interface/FrameXML/ChatFrame.lua
+++ b/Interface/FrameXML/ChatFrame.lua
@@ -1,6 +1,8 @@
 
 local forceinsecure = forceinsecure;
 local issecure = issecure;
+local scrub = scrub;
+local securecallfunction = securecallfunction;
 
 MESSAGE_SCROLLBUTTON_INITIAL_DELAY = 0;
 MESSAGE_SCROLLBUTTON_SCROLL_DELAY = 0.05;
@@ -5004,7 +5006,7 @@ function ChatEdit_SecureTabPressed(self)
 	local cmdString = lastTabComplete;
 	repeat	--The outer loop lets us go through multiple hash tables of commands.
 		repeat	--Loop through this table to find matching items.
-			cmdString = next(tabCompleteTables[self.tabCompleteTableIndex], cmdString);
+			cmdString = scrub(securecallfunction(next, tabCompleteTables[self.tabCompleteTableIndex], cmdString));
 		until ( not cmdString or strfind(cmdString, strupper(command), 1, 1) );	--Either we finished going through this table or we found a match.
 		if ( not cmdString ) then	--Nothing else in the current table, move to the next one.
 			self.tabCompleteTableIndex = self.tabCompleteTableIndex + 1;
```

The issue itself is that the SetText calls made by [ChatEdit_SecureTabPressed](https://github.com/Gethe/wow-ui-source/blob/0b5d4526f16f56a4c652d0d472620d6591a20f7d/Interface/FrameXML/ChatFrame.lua#L4981) store a tainted string. When you then press enter to execute the autocompleted command, GetText reads the tainted string and with it taints execution and then blocks execution of the command.

```
4/20 22:21:27.801  GetText()
4/20 22:21:27.801  An action was blocked because of taint from AddonUsage - Logout()
4/20 22:21:27.801      Interface/FrameXML/ChatFrame.lua:1703 ?()
4/20 22:21:27.801      Interface/FrameXML/ChatFrame.lua:5222
4/20 22:21:27.801      ChatEdit_ParseText()
4/20 22:21:27.801      Interface/FrameXML/ChatFrame.lua:4900 ChatEdit_SendText()
4/20 22:21:27.801      Interface/FrameXML/ChatFrame.lua:4936 ChatEdit_OnEnterPressed()
4/20 22:21:27.801      ChatFrame1EditBox:hatFrame.xml:133_OnEnterPressed()
```

There's two possible areas that can taint within ChatEdit_SecureTabPressed, however I believe only one of these is actually relevant.

The first is the [ChatFrame_ImportAllListsToHash](https://github.com/Gethe/wow-ui-source/blob/0b5d4526f16f56a4c652d0d472620d6591a20f7d/Interface/FrameXML/ChatFrame.lua#L2777) call. I believe this **shouldn't** be the issue here, as this function is called at multiple points in the chat frame infrastructure. By the time autocompletion actually occurs the lists it imports into the hash table will be completely empty, so there's no tainted entries within them to potentially traverse over.

The second more likely issue is instead the loop within ChatEdit_SecureTabPressed; this will **definitely** iterate over tainted entries in the `hash_ChatTypeInfoList` list (which is part of `tabCompleteTables`). As such this would then taint execution through to the SetText call at the end of the function, and with it break command execution.

**Comment by Meorawr** [comment]
Fixed in 10.2.7.
